hedera-node/docs/design/services/smart-contract-service/system-contract-abi-guidelines.md

```diff
@@ -0,0 +1,132 @@
+- When a method is to be available both for direct call by a contract and also available via
+  indirect call through a proxy (e.g., calling a Hedera Token type which thunks through a proxy)
+  then:
+  - The argument list types and order should match, the only difference being that the direct
+    call must pass the entity address as the first argument.  (It's supplied by the system 
+    implicitly in the proxy call case).
+  - In the case where the _proxy_ call is for an Ethereum ERC/EIP defined method (e.g., ERC-20
+    `approve`) then the proxy's method return type should match the Ethereum-defined method's
+    return type but the _direct_ call method return type should return a `ResponseCode` prepended
+    (if necessary) to the Ethereum-defined method's return type.
+    
+    E.g., from [HIP-514](https://hips.hedera.com/hip/hip-514) and [HIP-376](https://hips.hedera.com/hip/hip-376:
+    - Proxy signature:  `function approve(address spender, uint256 amount) external returns (bool)`
+    - Direct signature: `function approve(address token, address spender, uint256 amount) external returns (int64 responseCode)`
+      
+    Here the ERC-20 defined method simply returns a `bool`.  The direct call version returns
+    instead a `ResponseCode` and doesn't return a `bool` also, as the `ResponseCode` suffices
+    for success or failure.
+        
+## Notes
+
+### For the defining HIP
+
+The defining HIP should include a table listing:
+* method signature + returns
+  * _without_ argument names
+* method selector
+  * And after the HIP is approved: Add the selector to `4byte.directory`.
+* the method signature + returns with Solidity struct names if used for arguments
+  * The struct should _also_ be defined in the HIP
+* (optionally) the "expanded" method signature with structs expanded out to their fields, recursively,
+  as used to calculate the method selector
+* attributes (e.g., pure, view, etc)
+* checkmark on whether or not it is a proxy method
+* HAPI transaction(s) it exposes (if any)
+* reference to ERC/EIP (if any)
+
+And then each method should have a subsection defining it that starts with the method signature +
+returns, unexpanded, this time _with_ argument names.  This section should include more information
+on what the arguments actually represent (especially in the case of `string` or `bytes` arguments).
+
+There should also be a section listing new _response codes_ that the newly defined methods may
+return.
+* These are defined in a HIP becuase they must be added to the HAPI protobufs and thus they affect 
+  the permanent system state and record/block streams.
+  
+### For the smart contract repo
+
+In the smart contract repo each system contract is defined by an interface, and a helper/wrapper.
+For a system contract HFoo (e.g., HTS, HAS) there is a:
+- `IHFoo.sol` - an interface which defines the actual system contract ABI
+- `HFoo.sol` - a class that is a helper or wrapper over `IFoo.sol`, for the convenience of smart
```

Comment:
Perhaps this can be enhanced to include a description of IHRC interfaces when needed because the function will be called via a proxy contract.